pkg/activator/net/throttler.go

```diff
@@ -144,53 +135,51 @@ func newRevisionThrottler(revID types.NamespacedName,
 	}
 }
 
+func noop() {}
+
 // pickPod picks the first tracker that has open capacity if container concurrency
 // if limited, random pod otherwise.
-// TODO(vagabaov): make this locally ideal, rather than mostly ideal.
-func pickPod(tgs []*podTracker, cc int) *podTracker {
+func pickPod(ctx context.Context, tgs []*podTracker, cc int) (func(), *podTracker) {
 	// Infinite capacity, pick random. We have to do this
 	// otherwise _all_ the requests will go to the first pod
 	// since it has unlimited capacity.
 	if cc == 0 {
-		return tgs[rand.Intn(len(tgs))]
+		return noop, tgs[rand.Intn(len(tgs))]
 	}
 	for _, t := range tgs {
-		if t.HasCapacity() {
-			return t
+		if cb, ok := t.Reserve(ctx); ok {
+			return cb, t
 		}
 	}
 	// NB: as currently written this can never happen.
-	return nil
+	return nil, nil
```

Comment:
Should we remove a `nil` function pointer everywhere instead of a noop? We kinda need to guard against a nil anyway for this case.